File: C#/mk64cre.cs
if (args.Length < 2) {
	Console.WriteLine("Mario Kart 64 RAM Course Extractor\n");
	Console.WriteLine("Usage: mk64cre <input.bin> <output.obj>");
	return;
}

List<(short, short, short)> vertices = new();
List<(int, int, int)> triangles = new();

byte[] data = File.ReadAllBytes(args[0]);
Console.WriteLine($"Input file is {data.Length / 1024} KB");
Array.Resize(ref data, data.Length + 16);
for (int i = 1; i <= 16; i++)
	data[^i] = 0xFF;  // pad the file in case the vertex data table reaches end of file
int vertTable = ReadInt(data, 0x00150268);
Console.WriteLine($"Vertex       data starts at 0x{vertTable.ToString("X8")} according to pointer at 0x00150268");
int displayListTable = ReadInt(data, 0x00150274);
Console.WriteLine($"Display List data starts at 0x{displayListTable.ToString("X8")} according to pointer at 0x00150274");

// Read Vertex Table
int p;
for (p = vertTable; data[p+6] == 0 && data[p+7] >> 4 == 0; p += 16) {   //detect end of table by checking known 12 bits
	vertices.Add((
		ReadShort(data, p),
		ReadShort(data, p + 2),
		ReadShort(data, p + 4)
	));
}
Console.WriteLine($"Vertex Table ended at 0x{p.ToString("X8")}");

// Read DisplayList Table
int vertOffset = 0;
for (p = displayListTable; p < vertTable; p += 8) {
	switch (data[p]) {  // Switch on F3DEX commands
		case 0x04:  // Load Vertices
			vertOffset = ((data[p+5] << 16) | (data[p+6] << 8) | data[p+7]) >> 4;
			break;
		case 0xB1:  // Draw Two Triangles
			triangles.Add(((data[p+1] >> 1) + vertOffset, (data[p+2] >> 1) + vertOffset, (data[p+3] >> 1) + vertOffset));
			goto case 0xBF;
		case 0xBF:  // Draw One Triangle
			triangles.Add(((data[p+5] >> 1) + vertOffset, (data[p+6] >> 1) + vertOffset, (data[p+7] >> 1) + vertOffset));
			break;
	}
}

using (StreamWriter writer = new(args[1])) {
	vertices.ForEach(v => writer.WriteLine($"v {v.Item1} {v.Item2} {v.Item3}"));
	triangles.ForEach(t => writer.WriteLine($"f {t.Item1 + 1} {t.Item2 + 1} {t.Item3 + 1}"));
}
Console.WriteLine($"\nSaved {vertices.Count} vertices and {triangles.Count} triangles to {args[1]}");

static int ReadInt(byte[] data, int index) {
	return (data[index] << 24) | (data[index + 1] << 16) | (data[index + 2] << 8) | data[index + 3];
}

static short ReadShort(byte[] data, int index) {
	return (short)((data[index] << 8) | data[index + 1]);
}

File: C/mk64cre.c
#include <stdio.h>
#include <stdlib.h>
#include <inttypes.h>

int32_t readInt(uint8_t* data, int index) {
    return (data[index] << 24) | (data[index + 1] << 16) | (data[index + 2] << 8) | data[index + 3];
}

int16_t readShort(uint8_t* data, int index) {
    return (data[index] << 8) | data[index + 1];
}

int main(int argc, char** argv) {
    if (argc < 3) {
        printf("Mario Kart 64 RAM Course Extractor\n");
        printf("Usage: mk64cre <input.bin> <output.obj>\n");
        return 1;
    }

    FILE* inFile;
    errno_t err = fopen_s(&inFile, argv[1], "rb");
    if (err != 0) {
        printf("Error opening input file\n");
        return 1;
    }

    fseek(inFile, 0, SEEK_END);
    long fileSize = ftell(inFile);
    rewind(inFile);
    uint8_t* data = malloc(fileSize + 16);
    fread(data, 1, fileSize, inFile);
    fclose(inFile);
    memset(data + fileSize, 0xFF, 16);  // pad the file in case the vertex data table reaches end of file

    printf("Input file is %ld KB\n", fileSize / 1024);
    int vertTable = readInt(data, 0x00150268);
    printf("Vertex       data starts at 0x%08X according to pointer at 0x00150268\n", vertTable);
    int displayListTable = readInt(data, 0x00150274);
    printf("Display List data starts at 0x%08X according to pointer at 0x00150274\n", displayListTable);

    FILE* outFile;
    err = fopen_s(&outFile, argv[2], "w");
    if (err != 0) {
        printf("Error opening output file\n");
        return 1;
    }

    int vertCount = 0;
    int triCount = 0;
    int p;
    for (p = vertTable; data[p+6] == 0 && data[p+7] >> 4 == 0; p += 16) {   // detect table end by checking known 12 bits
        int16_t x = readShort(data, p);
        int16_t y = readShort(data, p + 2);
        int16_t z = readShort(data, p + 4);
        fprintf(outFile, "v %hd %hd %hd\n", x, y, z);
        vertCount++;
    }
    printf("Vertex Table ended at 0x%08X\n", p);

    int vertOffset = 0;
    for (p = displayListTable; p < vertTable; p += 8) {
        switch (data[p]) {
        case 0x04:  // Load Vertices
            vertOffset = ((data[p+5] << 16) | (data[p+6] << 8) | data[p+7]) >> 4;
            break;
        case 0xB1:  // Draw Two Triangles
            fprintf(outFile, "f %d %d %d\n", (data[p+1] >> 1) + vertOffset + 1, (data[p+2] >> 1) + vertOffset + 1, (data[p+3] >> 1) + vertOffset + 1);
            triCount++;
        case 0xBF:  // Draw One Triangle
            fprintf(outFile, "f %d %d %d\n", (data[p+5] >> 1) + vertOffset + 1, (data[p+6] >> 1) + vertOffset + 1, (data[p+7] >> 1) + vertOffset + 1);
            triCount++;
            break;
        }
    }
    fclose(outFile);
    free(data);
    printf("\nSaved %d vertices and %d triangles to %s\n", vertCount, triCount, argv[2]);
    return 0;
}
